KAFKA-13861: Fix the validateOnly behavior for CreatePartitions requests in KRaft mode

akhileshchg opened the pull request:
KAFKA-13861: Fix the validateOnly behavior for `CreatePartitions` requests in KRaft mode

The KRaft implementation of the `CreatePartitions` ignores the `validateOnly` flag in the
request and creates the partitions if the validations are successful. Fixed the behavior
not to create partitions upon validation if the `validateOnly` flag is true.

Add ControllerApi and ReplicationControlManager tests to verify the correct behavior.

akhileshchg committed:
KAFKA-13861: Fix the validateOnly behavior for CreatePartitions requests in KRaft mode

The KRaft implementation of the CreatePartitions ignores the validateOnly flag in the
request and creates the partitions if the validations are successful. Fixed the behavior
to make sure the flag is honored and the partitions are not actually created upon validation.

dengziming commented on core/src/main/scala/kafka/server/ControllerApis.scala:
```diff
@@ -767,7 +767,7 @@ class ControllerApis(val requestChannel: RequestChannel,
       }
     }
-    controller.createPartitions(context, topics).thenApply { results =>
+    controller.createPartitions(context, topics, request.validateOnly()).thenApply { results =>
```
Good catch, it seems we don't have a test case for this problem, to cover this we should change `PlaintextAdminIntegrationTest.testCreatePartitions` to support KRaft mode.

akhileshchg replied:
I enabled the test for both KRaft and Zk modes now. The existing test is robust and covers all the cases.

dengziming reviewed (changes requested):
I haven't gone through all the codes in detail, but we should first improve test case for it.

hachikuji commented on core/src/test/java/kafka/test/MockController.java:
```diff
@@ -410,10 +410,16 @@ public CompletableFuture<UpdateFeaturesResponseData> updateFeatures(
         throw new UnsupportedOperationException();
     }
 
+    boolean lastCreatePartitionsValidateOnly = false;
```
This is a little strange. Would it be enough to check whether the partition creation happened or not?

akhileshchg replied:
The `MockController` doesn't maintain any state of the topics/partitions that are created and deleted. So, I thought this was a more straightforward fix to check if the `validateOnly` flag is used. Please let met know if you have some other ideas.

akhileshchg replied:
Added a new unit test and remove `lastCreatePartitionsValidateOnly`

hachikuji commented on metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java:
```diff
@@ -1364,7 +1364,25 @@ Boolean isBrokerUnfenced(int brokerId) {
                 setErrorMessage(apiError.message()));
         }
-        return new ControllerResult<>(records, results, true);
+        StringBuilder resultsBuilder = new StringBuilder();
```
nit: would it be enough to print out the `results` directly. I think `CreatePartitionsTopicResult` already gives us a reasonable `toString`.

akhileshchg replied:
Sure. That seems reasonable. I did this to match the `createTopics` log statement.

hachikuji commented on metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java:
```diff
@@ -1364,7 +1364,25 @@ Boolean isBrokerUnfenced(int brokerId) {
+            resultsPrefix = ", ";
+        }
+        if (validateOnly) {
+            log.info("Validate-only CreatePartitions result(s): {}", resultsBuilder);
```
For `validateOnly`, maybe debug is good enough?

hachikuji reviewed (commented):
Thanks, good find. Left a few small comments.

divijvaidya commented on core/src/main/scala/kafka/server/ControllerApis.scala:
```diff
@@ -767,7 +767,7 @@ class ControllerApis(val requestChannel: RequestChannel,
       }
     }
-    controller.createPartitions(context, topics).thenApply { results =>
+    controller.createPartitions(context, topics, request.validateOnly()).thenApply { results =>
```
nit

you don't need the parenthesis in scala here. Simply `request.validateOnly` would work.

akhileshchg replied:
Done.

divijvaidya commented on metadata/src/main/java/org/apache/kafka/controller/QuorumController.java:
```diff
@@ -1664,13 +1664,14 @@ public CompletableFuture<UpdateFeaturesResponseData> updateFeatures(
     @Override
     public CompletableFuture<List<CreatePartitionsTopicResult>> createPartitions(
```
You can handle the `validateOnly` here and downstream function calls do not have to know about it. This would greatly simplify the code because now `ReplicationControl` does not have to deal with parsing the result and handling the `validateOnly` flag

e.g. this function implementation would change to
```
        if (topics.isEmpty()) {
            return CompletableFuture.completedFuture(Collections.emptyList());
        }

        return appendWriteEvent("createPartitions", context.deadlineNs(), () -> {
            final ControllerResult<List<CreatePartitionsTopicResult>> result = replicationControl.createPartitions(topics);
            return validateOnly ? result.withoutRecords() : result;
        });
```

akhileshchg replied:
That's a good idea. Will do this.

divijvaidya reviewed (changes requested):
Please correct me if I am wrong here but none of the tests added in the PR would fail if we have a bug where we don't set records as `empty`. Because the tests rely on mocks. Perhaps, we need an integration test where we need to assert whether the partition was actually created or not.

akhileshchg committed:
Address comments.
1. Enabled integration test for createPartitions in KRaft mode.
2. Fixed other comments in the PR.

akhileshchg committed:
Remove unncessary changes

akhileshchg reviewed (commented):
Addressed comments. Please take another look.

akhileshchg committed:
Add unit test case for validateOnly createPartitions call

hachikuji commented on core/src/test/scala/unit/kafka/server/ControllerApisTest.scala:
```diff
@@ -730,8 +729,45 @@ class ControllerApisTest {
+  }
+
+  @Test
+  def testValidateOnlyCreatePartitionsRequest(): Unit = {
```
Is this test basically the same as `testCreatePartitionsRequest`? Maybe we can get rid of one and turn the other into a `@ParameterizedTest` with `validateOnly` as the parameter?

hachikuji commented on core/src/test/scala/integration/kafka/api/PlaintextAdminIntegrationTest.scala:
```diff
@@ -499,23 +504,28 @@ class PlaintextAdminIntegrationTest extends BaseAdminIntegrationTest {
+      exceptionMsgStr = if (TestInfoUtils.isKRaft(testInfo)) {
+        "Topic already has 3 partition(s)."
+      } else {
+        "Topic already has 3 partitions."
```
I wonder if we can unify the error messages? The differences do not seem interesting.

akhileshchg replied:
There are too many error messages in the `ReplicationControl` layer, and I'm not sure how many other layers need to be changed. I think this can be a different PR to keep the messages intact if required.

hachikuji replied:
Fair enough. I don't think asserting the exact messages is all that useful to be honest, but not need to change that here.

hachikuji commented on metadata/src/main/java/org/apache/kafka/controller/Controller.java:
```diff
@@ -328,11 +328,15 @@ CompletableFuture<UpdateFeaturesResponseData> updateFeatures(
      * Create partitions on certain topics.
      *
      * @param topics        The list of topics to create partitions for.
+     * @param validateOnly  If true, create partitions is just validated and returns response
```
nit: how about this?
> If true, the request is validated, but no partitions will be created.

hachikuji commented on metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java:
```diff
@@ -1364,7 +1364,8 @@ Boolean isBrokerUnfenced(int brokerId) {
                 setErrorMessage(apiError.message()));
         }
-        return new ControllerResult<>(records, results, true);
+        log.debug("CreatePartitions result(s): {}", results);
```
Hmm.. It is useful to know in the logs if `validateOnly` was set.

akhileshchg replied:
I'll put a debug log in the controller layer before processing the final list of topics.

hachikuji replied:
I can still see this causing confusion. Perhaps we can move it to `QuorumController` where we have `validateOnly` and can make the message clear?

akhileshchg committed:
Address Jason's comments

akhileshchg committed:
minor changes

divijvaidya commented:
The code looks good but as I said earlier, we are missing tests that assert that `validateOnly` flag is working as expected. The tests added in this PR are simply validating whether the flag was passed or not. Instead, we want to validate that when the flag is set to `false` no change takes place.

But I think adding that test is out of scope for this PR since it may require refactoring and having a consistent boundary for termination of `validateOnly` flag for various APIs.

LGTM.

akhileshchg commented:
> The code looks good but as I said earlier, we are missing tests that assert that `validateOnly` flag is working as expected. The tests added in this PR are simply validating whether the flag was passed or not. Instead, we want to validate that when the flag is set to `false` no change takes place.
> 
> But I think adding that test is out of scope for this PR since it may require refactoring and having a consistent boundary for termination of `validateOnly` flag for various APIs.
> 
> LGTM.

@divijvaidya I enabled an integration test that tests the validateOnly field for createPartitions request for KRaft mode.

hachikuji commented on core/src/test/scala/integration/kafka/api/PlaintextAdminIntegrationTest.scala:
```diff
@@ -614,7 +669,12 @@ class PlaintextAdminIntegrationTest extends BaseAdminIntegrationTest {
+    assertEquals(exceptionMsgStr, e.getCause.getMessage)
     assertEquals(3, numPartitions(topic2))
 
     // finally, try to add partitions to a topic queued for deletion
```
It's probably worth adding a comment here. In KRaft, the topic is not "queued" for deletion in the same way that the  zookeeper controller does. The deletion occurs immediately which is why there is a difference in error codes below.

hachikuji reviewed (commented):
Thanks for the updates, just a couple more comments.

akhileshchg committed:
Address Jason's comments

hachikuji reviewed (approved):
Thanks, LGTM

dengziming commented on core/src/test/scala/integration/kafka/api/PlaintextAdminIntegrationTest.scala:
```diff
@@ -490,7 +490,12 @@ class PlaintextAdminIntegrationTest extends BaseAdminIntegrationTest {
         () => s"$desc: Expect InvalidPartitionsException when newCount is a decrease")
       assertTrue(e.getCause.isInstanceOf[InvalidPartitionsException], desc)
-      assertEquals("Topic currently has 3 partitions, which is higher than the requested 1.", e.getCause.getMessage, desc)
+      var exceptionMsgStr = if (TestInfoUtils.isKRaft(testInfo)) {
```
nit: We can directly use `isKRaftTest()` here, the below as well.

dengziming reviewed (approved):
LGTM, left a minor comment.

akhileshchg committed:
Address comments about usage of isKRaftTest()
